File: 0_Workflow and Pipelines CWICR/python/01-load-and-explore/load_parquet.py
"""
Load the DDC CWICR Parquet dataset and display basic information.

Uses the 100-row sample by default. Pass --full to load the complete
English dataset (900,225 rows).
"""

import argparse
import sys
from pathlib import Path

try:
    import pandas as pd
except ImportError:
    sys.exit("pandas is required. Install with: pip install pandas pyarrow")


SAMPLE_PATH = Path(__file__).resolve().parents[2] / "sample-data" / "sample_cwicr_en_100rows.parquet"
FULL_PATH = Path(__file__).resolve().parents[3] / "EN___DDC_CWICR" / "ENG_TORONTO_workitems_costs_resources_DDC_CWICR.parquet"


def main():
    parser = argparse.ArgumentParser(description="Load and explore DDC CWICR Parquet data.")
    parser.add_argument("--full", action="store_true", help="Load the full English dataset instead of the sample.")
    parser.add_argument("--path", type=str, default=None, help="Custom path to a Parquet file.")
    args = parser.parse_args()

    # Resolve file path
    if args.path:
        parquet_path = Path(args.path)
    elif args.full:
        parquet_path = FULL_PATH
    else:
        parquet_path = SAMPLE_PATH

    if not parquet_path.exists():
        sys.exit(f"File not found: {parquet_path}")

    print(f"Loading: {parquet_path}\n")
    df = pd.read_parquet(parquet_path)

    # Basic shape
    print(f"Rows: {len(df):,}")
    print(f"Columns: {len(df.columns)}\n")

    # Schema overview
    print("=== Column Names & Types ===")
    for col in df.columns:
        print(f"  {col:<50} {df[col].dtype}")
    print()

    # First 10 rows (selected columns for readability)
    display_cols = [
        "rate_code", "rate_original_name", "rate_unit",
        "total_cost_per_position", "department_name", "section_name",
    ]
    available = [c for c in display_cols if c in df.columns]
    print("=== First 10 Rows (key columns) ===")
    print(df[available].head(10).to_string(index=False))
    print()

    # Cost distribution
    if "total_cost_per_position" in df.columns:
        cost = df["total_cost_per_position"].dropna()
        print("=== total_cost_per_position Distribution ===")
        print(f"  Count:  {cost.count():,}")
        print(f"  Mean:   {cost.mean():,.2f} EUR")
        print(f"  Median: {cost.median():,.2f} EUR")
        print(f"  Min:    {cost.min():,.2f} EUR")
        print(f"  Max:    {cost.max():,.2f} EUR")
        print(f"  Std:    {cost.std():,.2f} EUR")
        print()

    # Department counts
    if "department_name" in df.columns:
        print("=== Rows per Department ===")
        dept_counts = df["department_name"].value_counts()
        for dept, count in dept_counts.items():
            print(f"  {dept:<50} {count:>8,}")
        print()

    print("Done.")


if __name__ == "__main__":
    main()

File: 0_Workflow and Pipelines CWICR/javascript/01-load-and-explore/load_parquet.mjs
/**
 * 01 - Load and Explore CWICR Parquet Data
 *
 * Demonstrates how to read the CWICR construction cost dataset from a Parquet
 * file using parquet-wasm and apache-arrow in Node.js.
 *
 * Usage:
 *   node load_parquet.mjs [path_to_parquet]
 */

import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { tableFromIPC } from "apache-arrow";
import { readParquet } from "parquet-wasm/node/arrow1.js";

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const __dirname = dirname(fileURLToPath(import.meta.url));
const defaultParquetPath = resolve(
  __dirname,
  "../../sample-data/sample_cwicr_en_100rows.parquet"
);
const parquetPath = process.argv[2] || defaultParquetPath;

// ---------------------------------------------------------------------------
// Load the Parquet file
// ---------------------------------------------------------------------------

console.log(`\nLoading Parquet file: ${parquetPath}\n`);

const parquetBytes = readFileSync(parquetPath);
const arrowIPCBytes = readParquet(new Uint8Array(parquetBytes));
const table = tableFromIPC(arrowIPCBytes);

// ---------------------------------------------------------------------------
// 1. Display the schema
// ---------------------------------------------------------------------------

console.log("=== Schema ===");
console.log(`Total columns: ${table.schema.fields.length}\n`);

for (const field of table.schema.fields) {
  console.log(`  ${field.name.padEnd(50)} ${field.type}`);
}

console.log(`\nTotal rows: ${table.numRows}\n`);

// ---------------------------------------------------------------------------
// 2. Preview the first 10 rows
// ---------------------------------------------------------------------------

console.log("=== First 10 Rows (selected columns) ===\n");

const previewColumns = [
  "rate_code",
  "rate_original_name",
  "rate_unit",
  "department_name",
  "total_cost_per_position",
];

// Build a simple table view
const rows = [];
for (let i = 0; i < Math.min(10, table.numRows); i++) {
  const row = {};
  for (const col of previewColumns) {
    const column = table.getChild(col);
    row[col] = column ? column.get(i) : "N/A";
  }
  rows.push(row);
}

console.table(rows);

// ---------------------------------------------------------------------------
// 3. Basic statistics for cost columns
// ---------------------------------------------------------------------------

console.log("\n=== Basic Statistics ===\n");

const costColumns = [
  "total_cost_per_position",
  "total_material_cost_per_position",
  "resource_price_per_unit_eur_current",
  "labor_hours_construction_workers",
];

for (const colName of costColumns) {
  const column = table.getChild(colName);
  if (!column) {
    console.log(`  Column '${colName}' not found — skipping.`);
    continue;
  }

  const values = [];
  for (let i = 0; i < column.length; i++) {
    const v = column.get(i);
    if (v !== null && v !== undefined && !Number.isNaN(Number(v))) {
      values.push(Number(v));
    }
  }

  if (values.length === 0) {
    console.log(`  ${colName}: no numeric values`);
    continue;
  }

  values.sort((a, b) => a - b);

  const count = values.length;
  const sum = values.reduce((a, b) => a + b, 0);
  const mean = sum / count;
  const min = values[0];
  const max = values[count - 1];
  const median =
    count % 2 === 0
      ? (values[count / 2 - 1] + values[count / 2]) / 2
      : values[Math.floor(count / 2)];

  console.log(`  ${colName}`);
  console.log(`    count  : ${count}`);
  console.log(`    min    : ${min.toFixed(2)}`);
  console.log(`    max    : ${max.toFixed(2)}`);
  console.log(`    mean   : ${mean.toFixed(2)}`);
  console.log(`    median : ${median.toFixed(2)}`);
  console.log();
}

console.log("Done.");
